### PR TITLE
Exchange refactoring

### PR DESCRIPTION
**Pull request by lmpn**
**event_bus.ex:**
- removed extra line to prevent warning from credo.

**matching_engine.ex:**
- corrected spread spec.
- handle_call for limit order and market order are now compliant with the order book's minimum and maximum price.

**order_book_test.exs:**

- added test for some edge cases and to test the increment_or_decrement function.

**order_book.ex:**
- added and refactored some specs.
- the price_time_match doesn't leave empty queues in the order book structure. (update_queue)
- in price_time_match when fetched order is matched as empty it is updated the max_bid or ask_min depending on the order and the order is queued.
- the calculation of min_ask or max_bid is compliant with the minimum and maximum price of the order book.
- price bound check added to increment_ask_min and decrement_bid_max.

### CODE REVIEW

File: apps/exchange/lib/exchange/order_book.ex [hunk 1]
```diff
@@ -277,12 +277,22 @@ defmodule Exchange.OrderBook do
...
+          |> Map.delete(price_point)
+        Map.put(order_book, side, updated_side_order_book)
+      _ ->
+        updated_side_order_book =
+          order_book
+          |> Map.fetch!(side)
+          |> Map.put(price_point, new_queue)
+
+        Map.put(order_book, side, updated_side_order_book)
+    end
```
Comment (emanuel) [inline apps/exchange/lib/exchange/order_book.ex:295]: we  could save a few lines without losing expressiveness doing:

```elixir
   ob  =  order_book |> Map.fetch!(side)

   updated_side_order_book =
   if Enum.count(new_queue) == 0 do
      ob |> Map.delete(price_point)
    else
      ob |> Map.put(price_point, new_queue)
   end

   Map.put(order_book, side, updated_side_order_book)
```

but your refactor is already a good improvement 💯

File: apps/exchange/lib/exchange/order_book.ex [hunk 2]
```diff
@@ -370,41 +384,49 @@ defmodule Exchange.OrderBook do
         !Enum.empty?(Map.get(order_book.buy, pp))
       end)
       |> List.first()
-
+    new_bid_max = case new_bid_max do
+      nil -> 1
```
Comment (emanuel) [inline apps/exchange/lib/exchange/order_book.ex:388]: good catch!

File: apps/exchange/lib/exchange/order_book.ex [hunk 3]
```diff
@@ -370,41 +384,49 @@ defmodule Exchange.OrderBook do
...
   end
 
   @doc """
   Updates the Order Book incrementing the ask_min by 1 or decrementing the bid_max by 1
-  taking into account the order's side
+  taking into account the order's side and price
   """
-  def increment_or_decrement(order_book, %Order{side: side}) do
+  @spec increment_or_decrement(order_book, Order.order()) :: order_book
+  def increment_or_decrement(order_book, %Order{side: side, price: price}) do
```
Comment (emanuel) [inline apps/exchange/lib/exchange/order_book.ex:399]: very good improvement !

File: apps/exchange/test/order_book_test.exs [hunk 4]
```diff
@@ -429,6 +428,32 @@ defmodule OrderBookTest do
       {:ok, %{order_book: sample_order_book(:AUXLND)}}
     end
 
+    test "increment_or_decrement", %{order_book: order_book} do
```
Comment (emanuel) [inline apps/exchange/test/order_book_test.exs:431]: good stuff !

**Comment by emanuel** [comment]
👍
